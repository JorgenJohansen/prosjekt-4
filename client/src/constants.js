const API_HOST_NAME = 'localhost';
const API_PORT = 8080;
const API_ROOT_PATH = '/api';

const getAPIAddress = () => {
    return `http://${API_HOST_NAME}:${API_PORT}${API_ROOT_PATH}`
};

const SEARCH_PAGE_OFFSET = 50;

// The first known motion picture in history was made in 1878, all our movies should be made that year or later
const MIN_SEARCH_YEAR = 1878;
const MAX_SEARCH_YEAR = 2100;

module.exports = {
    API_HOST_NAME,
    API_PORT,
    API_ROOT_PATH,
    getAPIAddress,
    SEARCH_PAGE_OFFSET,
    MIN_SEARCH_YEAR,
    MAX_SEARCH_YEAR
};
