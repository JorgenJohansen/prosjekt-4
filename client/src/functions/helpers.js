const checkForNA = (obj, value) => {
    if (value && value !== 'N/A') {
        return obj
    }
    return null
};

const checkForEmptyArr = (obj, arr) => {
    if(arr && arr !== []){
        return obj
    }
    return null
};

const stringifyArr = (arr) => {
    return arr.join(', ');
};

/**
 * Compares two objects and returns true if they are equal, and false otherwise.
 * Object parameters need to be ordered the same way for this function to be able to to compare them properly.
 * Should only be used for small objects, never on large result sets.
 * @param obj1 <Object>
 * @param obj2 <Object>
 * @return {boolean}
 */
const compareObjects = (obj1, obj2) => {
    return (JSON.stringify(obj1) === JSON.stringify(obj2))
};

module.exports = {
    checkForNA,
    checkForEmptyArr,
    stringifyArr,
    compareObjects
};
