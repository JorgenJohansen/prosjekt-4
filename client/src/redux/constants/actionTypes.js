export const ADD_RESULTS = 'ADD_RESULTS';
export const ADD_SORTING = 'ADD_SORTING';
export const ADD_SEARCH_ERROR = 'ADD_SEARCH_ERROR';
