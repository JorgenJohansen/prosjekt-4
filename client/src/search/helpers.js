import axios from "axios";


function handleResults(results, that){
    // Placeholder, venter pÃ¥ Redux-implementasjon
    that.setState({
        results: results
    })
}

function performSearch(searchText, that) {
    const requestString = `localhost:8080/movies/search?title=${searchText}`;
    console.log(requestString);

    axios.get(requestString).then(res => {
        console.log(res);
        console.log(res.data);
        handleResults(res.data, that);
    }).catch(error => {
        console.error(error);
    })
}

module.exports = {
    handleQuery:  (searchText, that) => {
        performSearch(searchText, that);
    }
};
