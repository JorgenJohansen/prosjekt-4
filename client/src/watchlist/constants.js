const WANT_TO_WATCH = 'wantToWatch';
const RM_WANT_TO_WATCH = 'rmWantToWatch';
const HAS_WATCHED = 'hasWatched';
const RM_HAS_WATCHED = 'rmHasWatched';

module.exports = {
    WANT_TO_WATCH,
    RM_WANT_TO_WATCH,
    HAS_WATCHED,
    RM_HAS_WATCHED,
};
