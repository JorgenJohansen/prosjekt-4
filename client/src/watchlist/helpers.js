/**
 * Retrieves the watchlist from LocalStorage. Returns empty object if it does not exist.
 * @return {*}
 */
const getWatchlistLocalStorage = () => {
    const result = JSON.parse(localStorage.getItem('watchList'));
    if ( result !== null) {
        return result;
    }
    return {}
};

/**
 * Writes data to watchlist in localStorage
 * @param data
 */
const writeWatchlistLocalStorage = data =>  {
    localStorage.setItem('watchList', JSON.stringify(data));
};

module.exports = {
  getWatchlistLocalStorage,
  writeWatchlistLocalStorage
};
