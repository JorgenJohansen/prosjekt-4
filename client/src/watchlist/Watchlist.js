/**
 * Component which shows lists of movies you 'want to watch' or 'has watched'.
 * The component uses localStorage, via a helper function i helpers.js, to get the imdbIDs of the movies, and is
 * therefore linked to the user via its browser. If no movies are saved to any of these lists, a message explaining how
 * movies can be added is showed.
 * The presentation of the lists is taken care of by MovieListContainer and movieListItem by passing them the imdbIDs.
 */

import React, { Component } from 'react';
import { Link } from "react-router-dom";
import { WANT_TO_WATCH, HAS_WATCHED } from './constants';
import MovieListContainer from "../common/movieList/MovieListContainer";
import { getWatchlistLocalStorage } from '../watchlist/helpers.js';
import './Watchlist.css';

class Watchlist extends Component {

    /**
     * Returns two arrays with movies from the watchlist in LocalStorage
     * @return {{hasWatched: Array, wantToWatch: Array}}
     */
    populateWatchlists = () => {
        let hasWatched = [];
        let wantToWatch = [];
        const watchlist = getWatchlistLocalStorage();

        for(let key of Object.keys(watchlist)) {
            // TODO: Check with API if movie exists in DB, remove it from watchlist if it does not exist
            if (watchlist[key] === HAS_WATCHED) {
                hasWatched.push({imdbID: key});
            } else if(watchlist[key] === WANT_TO_WATCH){
                wantToWatch.push({imdbID: key});
            }
        }
        return { hasWatched, wantToWatch };
    };

    render(){
        const watchlists = this.populateWatchlists();

        if (watchlists.wantToWatch.length === 0 && watchlists.hasWatched.length === 0) {
            return(
                <div className='watchlist'>
                    <h1>Du har ingen filmer i din Watchlist</h1>
                    <p> Du kan finne filmer på <Link to='/search/'>søkesiden,</Link> klikke deg inn på dem, og legge dem til i din Watchlist.
                    </p>
                </div>
            );
        } else {
            
            return(
                <div className='watchlist'>
                    {watchlists.wantToWatch.length !== 0 && <h1>Filmene du ønsker å se</h1> }
                    <MovieListContainer movies={watchlists.wantToWatch}/>
                    {watchlists.hasWatched.length !== 0 && <h1>Filmene du har sett</h1> }
                    <MovieListContainer movies={watchlists.hasWatched}/>
                </div>
            );
        }
    }
}

export default Watchlist;
