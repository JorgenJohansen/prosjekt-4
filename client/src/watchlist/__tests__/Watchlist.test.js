import React from 'react';
import renderer  from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import TestRenderer from 'react-test-renderer';
import { HAS_WATCHED, WANT_TO_WATCH } from '../constants';
import Watchlist from '../../watchlist/Watchlist';
import { MemoryRouter } from 'react-router-dom';

import { getWatchlistLocalStorage, writeWatchlistLocalStorage} from '../helpers';

describe('Rendering', () => {

    it('matches snaphot', () => {
        //Siden watchlist inneholder en link må vi wrappe den med MemoryRouter
        const tree = renderer.create(<MemoryRouter><Watchlist /></MemoryRouter>);
        expect(tree).toMatchSnapshot();
    })
});

//Generell localstoragemock som gir oss mulighet til å hente, legge til og slette elementer ved
//hjelp av metodene som er skrevet under, dette er også til hjelp med tanke på at vi akkurat nå
//ikke har funksjoner i helpers.js som kan slette elementer og fjerne alt fra localStorage
//med tanke på at dette krever ekstra funksjonalitet som vi ikke har tid til å implementere
//eller er nødvendig for siden vår
const localStorageMock = (function(){
    let store = {};
    return {
        getItem(key){
            return store[key]|| null;
        },
        setItem(key,value){
            store[key] = value.toString();
        },
        removeItem(key){
            delete store[key]
        },
        clear(){
            store = {};
        },
    };
})();

Object.defineProperty(window,'localStorage', {
    value: localStorageMock,
});

describe('teste localStorage mock', () => {
    beforeEach(() => localStorage.clear());

    it('localStorage er tomt når vi inititialiserer det', () => expect(localStorage.getItem("wantToWatch")).toBeNull());

    it('legge til en film til wantToWatch', () => {
        localStorage.setItem('wantToWatch','Batman');
        expect(localStorage.getItem('wantToWatch')).toEqual('Batman');
    });

    it('fjerne filmen som er lagt til i wantToWatch', () => {
        localStorage.removeItem('wantToWatch');
        const movie = localStorage.getItem('wantToWatch');
        expect(movie).toBeNull();
    });
});



describe('Teste watchlist localStorage', () => {
    beforeEach(() => localStorage.clear());

    it('legge til en film til watchlist', () => {
        const data = {wantToWatch:"batman"};
        writeWatchlistLocalStorage(data);
        const movie = getWatchlistLocalStorage();
        expect(movie).toEqual(data);
    });

    it('returnere et tomt objekt hvis det er ingenting i localStorage', () => {
        localStorage.clear();
        const movie = getWatchlistLocalStorage();
        expect(movie).toEqual({});
    });
});

describe('populateWatchLists', () => {
    beforeEach(() => localStorage.clear());
    it('should return empty lists', () => {

        const data = {};

        const watchLists = {
            wantToWatch: [],
            hasWatched: []
        };
        writeWatchlistLocalStorage(data);


        const watchList = mount(<MemoryRouter><Watchlist/></MemoryRouter>).find(Watchlist).instance();
        expect(watchList.populateWatchlists()).toEqual(watchLists)
    });

    it('should return empty hasWatched', () => {

        const data = {
            'tt0361763': WANT_TO_WATCH,
            'tt1194172': WANT_TO_WATCH
        };

        const watchLists = {
            wantToWatch: [
                {imdbID: 'tt0361763'},
                {imdbID: 'tt1194172'}
            ],
            hasWatched: []
        };
        writeWatchlistLocalStorage(data);


        const watchList = mount(<MemoryRouter><Watchlist/></MemoryRouter>).find(Watchlist).instance();
        expect(watchList.populateWatchlists()).toEqual(watchLists)
    });

    it('should return empty wantToWatch', () => {

        const data = {
            'tt0057875': HAS_WATCHED,
            'tt1228961': HAS_WATCHED,
            'tt0061386': HAS_WATCHED,
            'tt0103776': HAS_WATCHED
        };

        const watchLists = {
            wantToWatch: [],
            hasWatched: [
                {imdbID: 'tt0057875'},
                {imdbID: 'tt1228961'},
                {imdbID: 'tt0061386'},
                {imdbID: "tt0103776"}
            ]
        };
        writeWatchlistLocalStorage(data);

        const watchList = mount(<MemoryRouter><Watchlist/></MemoryRouter>).find(Watchlist).instance();
        expect(watchList.populateWatchlists()).toEqual(watchLists)
    });

    it('should return both arrays with content', () => {

        const data = {
            'tt0057875': HAS_WATCHED,
            'tt0361763': WANT_TO_WATCH,
            'tt1228961': HAS_WATCHED,
            'tt0061386': HAS_WATCHED,
            'tt1194172': WANT_TO_WATCH,
            'tt0103776': HAS_WATCHED
        };

        const watchLists = {
            wantToWatch: [
                {imdbID: 'tt0361763'},
                {imdbID: 'tt1194172'}
            ],
            hasWatched: [
                {imdbID: 'tt0057875'},
                {imdbID: 'tt1228961'},
                {imdbID: 'tt0061386'},
                {imdbID: "tt0103776"}
            ]
        };
        writeWatchlistLocalStorage(data);

        const watchList = mount(<MemoryRouter><Watchlist/></MemoryRouter>).find(Watchlist).instance();
        expect(watchList.populateWatchlists()).toEqual(watchLists)
    });


});



describe('rendering correctly with lists', () => {
    beforeEach(() => localStorage.clear());

    it('should return empty lists', () => {
        // Will render "du har ingen filmer i din watchlist..."
        const tree = renderer.create(<MemoryRouter><Watchlist /></MemoryRouter>);
        expect(tree).toMatchSnapshot();
    });

    it('should return empty hasWatched', () => {

        const data = {
            'tt0361763': WANT_TO_WATCH,
            'tt1194172': WANT_TO_WATCH
        };
        writeWatchlistLocalStorage(data);

        // Will render with "Vil se" header and one movie-list-container with two movie-list-items
        const tree = renderer.create(<MemoryRouter><Watchlist /></MemoryRouter>);
        expect(tree).toMatchSnapshot();
    });

    it('should return empty wantToWatch', () => {

        const data = {
            'tt0057875': HAS_WATCHED,
            'tt1228961': HAS_WATCHED,
            'tt0061386': HAS_WATCHED,
            'tt0103776': HAS_WATCHED
        };
        writeWatchlistLocalStorage(data);

        // Will render with "Har" header and one movie-list-container movie-list-container with two movie-list-items
        const tree = renderer.create(<MemoryRouter><Watchlist /></MemoryRouter>);
        expect(tree).toMatchSnapshot();
    });

    it('should return both arrays with content', () => {

        const data = {
            'tt0057875': HAS_WATCHED,
            'tt0361763': WANT_TO_WATCH,
            'tt1228961': HAS_WATCHED,
            'tt0061386': HAS_WATCHED,
            'tt1194172': WANT_TO_WATCH,
            'tt0103776': HAS_WATCHED
        };
        writeWatchlistLocalStorage(data);

        // Will render with "Vil se" and "Har sett" headers and their respective movie-list-container's,
        // both will have two movie-list-items
        const tree = renderer.create(<MemoryRouter><Watchlist /></MemoryRouter>);
        expect(tree).toMatchSnapshot();
    });

});

