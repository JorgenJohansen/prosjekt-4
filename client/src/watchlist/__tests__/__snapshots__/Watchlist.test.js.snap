// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering matches snaphot 1`] = `
<div
  className="watchlist"
>
  <h1>
    Du har ingen filmer i din Watchlist
  </h1>
  <p>
     Du kan finne filmer på 
    <a
      href="/search/"
      onClick={[Function]}
    >
      søkesiden,
    </a>
     klikke deg inn på dem, og legge dem til i din Watchlist.
  </p>
</div>
`;

exports[`rendering correctly with lists should return both arrays with content 1`] = `
<div
  className="watchlist"
>
  <h1>
    Filmene du ønsker å se
  </h1>
  <div
    className="movie-list-container"
  >
    <div>
      <div
        className="movie-list-item"
      >
        <a
          href="/movie/tt0361763"
          onClick={[Function]}
        >
          <div
            className="image-container no-image"
          />
          <p>
            Laster ...
          </p>
        </a>
      </div>
      <div
        className="movie-list-item"
      >
        <a
          href="/movie/tt1194172"
          onClick={[Function]}
        >
          <div
            className="image-container no-image"
          />
          <p>
            Laster ...
          </p>
        </a>
      </div>
    </div>
  </div>
  <h1>
    Filmene du har sett
  </h1>
  <div
    className="movie-list-container"
  >
    <div>
      <div
        className="movie-list-item"
      >
        <a
          href="/movie/tt0057875"
          onClick={[Function]}
        >
          <div
            className="image-container no-image"
          />
          <p>
            Laster ...
          </p>
        </a>
      </div>
      <div
        className="movie-list-item"
      >
        <a
          href="/movie/tt1228961"
          onClick={[Function]}
        >
          <div
            className="image-container no-image"
          />
          <p>
            Laster ...
          </p>
        </a>
      </div>
      <div
        className="movie-list-item"
      >
        <a
          href="/movie/tt0061386"
          onClick={[Function]}
        >
          <div
            className="image-container no-image"
          />
          <p>
            Laster ...
          </p>
        </a>
      </div>
      <div
        className="movie-list-item"
      >
        <a
          href="/movie/tt0103776"
          onClick={[Function]}
        >
          <div
            className="image-container no-image"
          />
          <p>
            Laster ...
          </p>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`rendering correctly with lists should return empty hasWatched 1`] = `
<div
  className="watchlist"
>
  <h1>
    Filmene du ønsker å se
  </h1>
  <div
    className="movie-list-container"
  >
    <div>
      <div
        className="movie-list-item"
      >
        <a
          href="/movie/tt0361763"
          onClick={[Function]}
        >
          <div
            className="image-container no-image"
          />
          <p>
            Laster ...
          </p>
        </a>
      </div>
      <div
        className="movie-list-item"
      >
        <a
          href="/movie/tt1194172"
          onClick={[Function]}
        >
          <div
            className="image-container no-image"
          />
          <p>
            Laster ...
          </p>
        </a>
      </div>
    </div>
  </div>
  <div
    className="movie-list-container"
  >
    <div />
  </div>
</div>
`;

exports[`rendering correctly with lists should return empty lists 1`] = `
<div
  className="watchlist"
>
  <h1>
    Du har ingen filmer i din Watchlist
  </h1>
  <p>
     Du kan finne filmer på 
    <a
      href="/search/"
      onClick={[Function]}
    >
      søkesiden,
    </a>
     klikke deg inn på dem, og legge dem til i din Watchlist.
  </p>
</div>
`;

exports[`rendering correctly with lists should return empty wantToWatch 1`] = `
<div
  className="watchlist"
>
  <div
    className="movie-list-container"
  >
    <div />
  </div>
  <h1>
    Filmene du har sett
  </h1>
  <div
    className="movie-list-container"
  >
    <div>
      <div
        className="movie-list-item"
      >
        <a
          href="/movie/tt0057875"
          onClick={[Function]}
        >
          <div
            className="image-container no-image"
          />
          <p>
            Laster ...
          </p>
        </a>
      </div>
      <div
        className="movie-list-item"
      >
        <a
          href="/movie/tt1228961"
          onClick={[Function]}
        >
          <div
            className="image-container no-image"
          />
          <p>
            Laster ...
          </p>
        </a>
      </div>
      <div
        className="movie-list-item"
      >
        <a
          href="/movie/tt0061386"
          onClick={[Function]}
        >
          <div
            className="image-container no-image"
          />
          <p>
            Laster ...
          </p>
        </a>
      </div>
      <div
        className="movie-list-item"
      >
        <a
          href="/movie/tt0103776"
          onClick={[Function]}
        >
          <div
            className="image-container no-image"
          />
          <p>
            Laster ...
          </p>
        </a>
      </div>
    </div>
  </div>
</div>
`;
